### PR TITLE
Upgrade infer commands from APP_VERSION

### PR DESCRIPTION
**Pull request by prastoin**
# Introduction
This PR refactors the way we previously manually handled the upgrade command `versionTo` and `versionFrom` values to be replaced by a programmatic inferring using the `APP_VERSION` env variable. It raises new invariant edge cases that are covered by new tests and so on

Please keep in mind that an upgrade will run agnostically of any `patch` semver value as it should be done only when releasing a `major/minor` version update
[Related discord thread](https://discord.com/channels/1130383047699738754/1368953221921505280)

## Testing in local
In order to test in local we have to define an `APP_VERSION` value in `packages/twenty-server/.env` following semver ( or not 🙃 )

## Logs example
```ts
Computing new Datasource for cacheKey: 20202020-1c25-4d02-bf25-6aeccf7ea419-8 out of 0
query: SELECT * FROM current_schema()
query: SELECT version();
[Nest] 37872  - 05/06/2025, 4:07:21 PM     LOG [UpgradeCommand] Initialized upgrade context with:
   - currentVersion (migrating to): 0.53.0
   - fromWorkspaceVersion: 0.52.0
   - 2 commands
[Nest] 37872  - 05/06/2025, 4:07:21 PM     LOG [UpgradeCommand] Upgrading workspace 20202020-1c25-4d02-bf25-6aeccf7ea419 from=0.52.0 to=0.53.0 1/2
[Nest] 37872  - 05/06/2025, 4:07:21 PM     LOG [UpgradeCommand] Upgrade for workspace 20202020-1c25-4d02-bf25-6aeccf7ea419 ignored as is already at a higher version.
[Nest] 37872  - 05/06/2025, 4:07:21 PM     LOG [UpgradeCommand] Running command on workspace 3b8e6458-5fc1-4e63-8563-008ccddaa6db 2/2
Computing new Datasource for cacheKey: 3b8e6458-5fc1-4e63-8563-008ccddaa6db-8 out of 0
query: SELECT * FROM current_schema()
query: SELECT version();
[Nest] 37872  - 05/06/2025, 4:07:21 PM     LOG [UpgradeCommand] Upgrading workspace 3b8e6458-5fc1-4e63-8563-008ccddaa6db from=0.52.0 to=0.53.0 2/2
[Nest] 37872  - 05/06/2025, 4:07:21 PM     LOG [UpgradeCommand] Upgrade for workspace 3b8e6458-5fc1-4e63-8563-008ccddaa6db ignored as is already at a higher version.
[Nest] 37872  - 05/06/2025, 4:07:21 PM     LOG [UpgradeCommand] Command completed!
```

## Misc
Related to https://github.com/twentyhq/twenty/issues/11780

### CODE REVIEW

File: packages/twenty-server/src/database/commands/command-runners/upgrade.command-runner.ts [hunk 1]
```diff
@@ -124,10 +180,15 @@ export abstract class UpgradeCommandRunner extends ActiveOrSuspendedWorkspacesMi
...
-  protected abstract runBeforeSyncMetadata(
-    args: RunOnWorkspaceArgs,
-  ): Promise<void>;
-  protected abstract runAfterSyncMetadata(
-    args: RunOnWorkspaceArgs,
-  ): Promise<void>;
+  public readonly runBeforeSyncMetadata = async (args: RunOnWorkspaceArgs) => {
+    for (const command of this.commands.beforeSyncMetadata) {
+      await command.runOnWorkspace(args);
+    }
```
[the remark this answers is not in the document]
Comment (prastoin) [reply packages/twenty-server/src/database/commands/command-runners/upgrade.command-runner.ts:186]: But used the arrow function readonly trick to avoid any mutation

File: packages/twenty-server/src/database/commands/command-runners/__tests__/upgrade.command-runner.spec.ts [hunk 2]
```diff
@@ -12,44 +11,47 @@ import { Workspace } from 'src/engine/core-modules/workspace/workspace.entity';
...
-  }
+  allCommands = {
+    '1.0.0': {
+      beforeSyncMetadata: [],
+      afterSyncMetadata: [],
+    },
+    invalid: {
+      beforeSyncMetadata: [],
+      afterSyncMetadata: [],
+    },
```
[the remark this answers is not in the document]
Comment (prastoin) [reply packages/twenty-server/src/database/commands/command-runners/__tests__/upgrade.command-runner.spec.ts:36]: Nice one

[the remark this answers is not in the document]
Comment (prastoin) [reply packages/twenty-server/src/database/commands/command-runners/__tests__/upgrade.command-runner.spec.ts:36]: Good bot

File: packages/twenty-server/src/database/commands/command-runners/upgrade.command-runner.ts [hunk 3]
```diff
@@ -33,15 +42,62 @@ export abstract class UpgradeCommandRunner extends ActiveOrSuspendedWorkspacesMi
...
+      this.currentAppVersion,
+      this.commands,
+      this.fromWorkspaceVersion,
+    ].every(isDefined);
+
+    if (ugpradeContextAlreadyDefined) {
+      return;
+    }
+
+    const currentAppVersion = this.retrieveToVersionFromAppVersion();
```
Comment (ijreilly) [inline packages/twenty-server/src/database/commands/command-runners/upgrade.command-runner.ts:56]: I don't really understand the name of that method, why is not just retrieveAppVersion() ?

File: packages/twenty-server/src/database/commands/command-runners/upgrade.command-runner.ts [hunk 4]
```diff
@@ -33,15 +42,62 @@ export abstract class UpgradeCommandRunner extends ActiveOrSuspendedWorkspacesMi
...
+
+    const currentCommands = this.allCommands[currentAppVersion];
+
+    if (!isDefined(currentCommands)) {
+      throw new Error(
+        `No commands found for version ${currentAppVersion}. Please check the commands record.`,
```
Comment (ijreilly) [inline packages/twenty-server/src/database/commands/command-runners/upgrade.command-runner.ts:62]:
```suggestion
        `No command found for version ${currentAppVersion}. Please check the commands record.`,
```

File: packages/twenty-server/src/database/commands/command-runners/upgrade.command-runner.ts [hunk 5]
```diff
@@ -33,15 +42,62 @@ export abstract class UpgradeCommandRunner extends ActiveOrSuspendedWorkspacesMi
...
+      );
+    }
+
+    const allCommandsKeys = Object.keys(this.allCommands);
```
Comment (ijreilly) [inline packages/twenty-server/src/database/commands/command-runners/upgrade.command-runner.ts:66]:
```suggestion
    const commandsVersions = Object.keys(this.allCommands);
```

File: packages/twenty-server/src/database/commands/command-runners/upgrade.command-runner.ts [hunk 6]
```diff
@@ -33,15 +42,62 @@ export abstract class UpgradeCommandRunner extends ActiveOrSuspendedWorkspacesMi
...
+    const previousVersion = getPreviousVersion({
```
Comment (ijreilly) [inline packages/twenty-server/src/database/commands/command-runners/upgrade.command-runner.ts:67]:
```suggestion
    const previousVersion = getPreviousAppVersion({
```

File: packages/twenty-server/src/database/commands/command-runners/upgrade.command-runner.ts [hunk 7]
```diff
@@ -33,15 +42,62 @@ export abstract class UpgradeCommandRunner extends ActiveOrSuspendedWorkspacesMi
...
+      currentVersion: currentAppVersion,
+      versions: allCommandsKeys,
+    });
+
+    if (!isDefined(previousVersion)) {
+      throw new Error(
+        `No previous version found for version ${currentAppVersion}. Please check the commands record available ${allCommandsKeys.join(', ')}.`,
```
Comment (ijreilly) [inline packages/twenty-server/src/database/commands/command-runners/upgrade.command-runner.ts:74]: what's a command record ? I don't understand what we are supposed to check if this happens

File: packages/twenty-server/src/database/commands/command-runners/upgrade.command-runner.ts [hunk 8]
```diff
@@ -33,15 +42,62 @@ export abstract class UpgradeCommandRunner extends ActiveOrSuspendedWorkspacesMi
     super(workspaceRepository, twentyORMGlobalManager);
   }
 
+  private computeFromToVersionAndCommandsToRunForCurrentAppVersion() {
```
Comment (ijreilly) [inline packages/twenty-server/src/database/commands/command-runners/upgrade.command-runner.ts:45]:
```suggestion
  private setFromToVersionAndCommandsToRunForCurrentAppVersion() {
```

File: packages/twenty-server/src/utils/version/__tests__/get-previous-version.spec.ts [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+
+import { getPreviousVersion } from 'src/utils/version/get-previous-version';
+
+type GetPreviousVersionTestCase = EachTestingContext<{
+  versions: string[];
+  currentVersion: string;
+  expected: string | undefined;
+}>;
+
+describe('get-previous-version', () => {
```
Comment (ijreilly) [inline packages/twenty-server/src/utils/version/__tests__/get-previous-version.spec.ts:11]: great to have this tested! 
to remain consistent with our other tests we can use the classic describe(() => { ... it('should order a Basic version sequence", ... }

but that's a nit / up to you

**Review by ijreilly — COMMENTED** [review]
In progress - still need to test !

File: packages/twenty-server/src/database/commands/command-runners/upgrade.command-runner.ts (hunk 6 above)
Comment (prastoin) [reply packages/twenty-server/src/database/commands/command-runners/upgrade.command-runner.ts:67]: `getPreviousVersion` is a util agnostical of the passed args source

File: packages/twenty-server/src/database/commands/command-runners/upgrade.command-runner.ts (hunk 7 above)
Comment (prastoin) [reply packages/twenty-server/src/database/commands/command-runners/upgrade.command-runner.ts:74]: About to add the record's var name ( allCommands )

File: packages/twenty-server/src/database/commands/command-runners/upgrade.command-runner.ts (hunk 5 above)
Comment (prastoin) [reply packages/twenty-server/src/database/commands/command-runners/upgrade.command-runner.ts:66]: Could be misleading as `commands` also exists

File: packages/twenty-server/src/database/commands/command-runners/upgrade.command-runner.ts (hunk 3 above)
Comment (prastoin) [reply packages/twenty-server/src/database/commands/command-runners/upgrade.command-runner.ts:56]: This is legacy will rename

File: packages/twenty-server/src/database/commands/command-runners/upgrade.command-runner.ts (hunk 7 above)
Comment (prastoin) [reply packages/twenty-server/src/database/commands/command-runners/upgrade.command-runner.ts:74]:
```ts
    this.allCommands = {
      '0.43.0': commands_043,
      '0.44.0': commands_044,
      '0.50.0': commands_050,
      '0.51.0': commands_051,
      '0.52.0': commands_052,
      '0.53.0': commands_053,
    };
```

**Review by ijreilly — APPROVED** [review]

File: packages/twenty-server/src/database/commands/command-runners/upgrade.command-runner.ts (hunk 8 above)
Comment (prastoin) [reply packages/twenty-server/src/database/commands/command-runners/upgrade.command-runner.ts]: In fact yep renamed from `computeFromToVersionAndCommandsToRunForCurrentAppVersion` to `setUpgradeContextVersionsAndCommandsForCurrentAppVersion`

File: packages/twenty-server/src/database/commands/command-runners/upgrade.command-runner.ts (hunk 3 above)
Comment (prastoin) [reply packages/twenty-server/src/database/commands/command-runners/upgrade.command-runner.ts]: Already renamed to `retrieveCurrentAppVersion` [2408255](https://github.com/twentyhq/twenty/pull/11881/commits/240825508d0f5766ffdb1140568e780fbc8712e6)
